### PR TITLE
[Webpack4] Migration to Webpack4, Babel 7, Node 10.15.3 & npm 6.4.1

### PR DESCRIPTION
**Pull request by MaxwellKendall**
Recommend merging #1074 into DEV after verifying these changes in Sandbox. This will put Webpack4, Babel7, React16, & Redux4 into DEV.

**High level description:**
This PR will update the project to use webpack 4 for front end bundling. The intent is to create a bundle equally or more performant than the previous bundle.

**Technical details:**
Main difficulty in migration was around removing previous configuration made obsolete by webpack4 defaults and utilizing the new configuration properties and new loaders. Also, was quite a pain updating to babel 7. Sorry for the 50+ commits. Would love to discuss the config and possibly iterate to provide more comments in the config explaining the choices made.

#### Engines: Node 10.15.3 & npm 6.4.1

**Existing Production Bundling Output from DEV branch using Webpack 2:**
image
-  7 images are also too big
- All JS files are too big
- All CSS files are too big

**New Production Bundling Output from this branch using Webpack4:**
image
- 5 images are too big
- Relatively fewer JS files are too big
- CSS file is too big but smaller than previously 

**JIRA Ticket:**
[DEV-1234](https://federal-spending-transparency.atlassian.net/browse/DEV-1234)

The following are ALL required for the PR to be merged:
- [x] Code review
- [x] Make `npm run dev` script point at the `dev` webpack config again
- [x] Ensured new production build is more or equally performant

### CODE REVIEW

**Comment by MaxwellKendall** [comment]
**Current progress:** 

* Development config:
  - Successfully spin up the DEV server using `npm run start`.
  - With the current configuration, images not loaded via CSS cannot be found. 
    - With the current configuration, we would need to change the way we reference the `src` and `srcSet` of images in our react components. For an example of how to reference image urls w/in a react component with the current configuration check out the `SpendingExplorer` component.

* Production Config:
  - Have not started

* Build Time / Bundle Size Analysis
  - Have not started

CC: @ebdabbs & @jonhill13

File: src/js/app.jsx [hunk 1]
```diff
@@ -1,22 +1,25 @@
 import React from 'react';
-import Perf from 'react-addons-perf';
 import { render } from 'react-dom';
```
Comment (MaxwellKendall) [inline src/js/app.jsx]: This change should have been in #1074.

See react-documentation about this pacakge being no-longer supported effective w/ react 16: https://reactjs.org/docs/perf.html

File: src/js/app.jsx [hunk 2]
```diff
@@ -1,22 +1,25 @@
 import React from 'react';
-import Perf from 'react-addons-perf';
```
Comment (MaxwellKendall) [inline src/js/app.jsx]: This change should have been in #1074.

See react-documentation about this pacakge being no-longer supported effective w/ react 16: https://reactjs.org/docs/perf.html

File: babel.config.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+const presets = [
+    [
+        "@babel/preset-env",
+        {
+            targets: {
+                ie: "10",
+                node: "current"
```
Comment (ebdabbs) [inline babel.config.js]: Does this always point to the latest version of node as it is released? Should we specify 10?

File: jest.config.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+module.exports = {
+    rootDir: ".",
+    testRegex: "tests/.*-(test)\\.jsx?$",
+    verbose: true,
+    bail: false,
+    collectCoverage: true,
+    collectCoverageFrom: [
+        "src/js/containers/**/*.{js,jsx}",
+        "src/js/redux/reducers/**/*.{js,jsx}",
+        "!node_modules/**",
```
Comment (ebdabbs) [inline jest.config.js]: When running `npm test` I'm getting this output which I haven't seen before:
```Running coverage on untested files...  console.error node_modules/@fortawesome/react-fontawesome/index.js:325
    Could not find icon { prefix: 'fas', iconName: 'chart-pie' }

  console.error node_modules/@fortawesome/react-fontawesome/index.js:325
    Could not find icon { prefix: 'fas', iconName: 'chart-pie' }

  console.error node_modules/@fortawesome/react-fontawesome/index.js:325
    Could not find icon { prefix: 'fas', iconName: 'chart-pie' }

  console.error node_modules/@fortawesome/react-fontawesome/index.js:325
    Could not find icon { prefix: 'fas', iconName: 'chart-pie' }

  console.error node_modules/@fortawesome/react-fontawesome/index.js:325
    Could not find icon { prefix: 'fas', iconName: 'chart-pie' }
```
wondering if we need to mock the fontawesome library or something.

File: webpack/webpack.prod.config.js [hunk 5]
```diff
@@ -1,25 +1,51 @@
...
-const webpack = require('webpack');
 const merge = require('webpack-merge');
-const ExtractTextPlugin = require('extract-text-webpack-plugin');
-const ParallelUglifyPlugin = require('webpack-parallel-uglify-plugin');
-const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
+const webpack = require('webpack');
+const OptimizeCssAssetsPlugin = require("optimize-css-assets-webpack-plugin");
+const MiniCssExtractPlugin = require('mini-css-extract-plugin');
+const UglifyJsPlugin = require("uglifyjs-webpack-plugin");
+// const BundleAnalyzer = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
```
Comment (ebdabbs) [inline webpack/webpack.prod.config.js]: Can this line (and 49) be removed?

File: Dockerfile [hunk 6]
```diff
@@ -1,4 +1,4 @@
-FROM node:6
+FROM node:10.15.3
```
[the remark this answers is not in the document]
Comment (MaxwellKendall) [reply Dockerfile]: I think so. This is the latest long-term stable version. I'm not sure what just putting 10 would use?

File: jest.config.js (hunk 4 above)
Comment (MaxwellKendall) [reply jest.config.js]: @ebdabbs -- 

All the configuration is the same, but these errors don't appear in the `dev` branch.  Also, there are some other errors from tests being executed that points to async errors. I was able to fix some of them by adding `async/await` keywords to some of the tests.

All of this I think is likely attributable to the upgrade to `Jest`.

**Comment by MaxwellKendall** [comment]
> accountLandingPageMetaTags

1. I think the error logging in webpack4 is perhaps a bit better and that's why we're seeing this now? It looks like this is incorrectly being imported on the `dev` branch as well. If we were to address this within the scope of this PR, would we just remove the import statement?
2. I removed the document!

File: jest.config.js (hunk 4 above)
Comment (ebdabbs) [reply jest.config.js]: Agreed, I think it falls out of the scope of this PR to address this!

**Pull request merged**